StateMachine: Merge Commit and Transfer

koekiebox opened the pull request:
Please see:
https://github.com/coilhq/tigerbeetle/issues/68

koekiebox committed:
Issue #63:
* Initial commit for Transfer/Commit merge.
* Backup

koekiebox committed:
Issue #63:
* Removed CreateCommit and CreateCommitResult

koekiebox committed:
Issue #63:
* Update benchmark.zig

koekiebox committed:
Updated state_machine and vsr.zig.

koekiebox committed:
Merge branch 'main' into issues/68-merge-commit-transfer

koekiebox committed:
Make use of pending vs posting flag for Transfer. Fix padding on the TransferFlags.

koekiebox committed:
Remove the [preimage] and [condition] flags with [hashlock]

koekiebox committed:
Validate for duplicate transfer prior to exceeding debits/credits verification.

koekiebox committed:
Fix test cases and ensure all is passing.

koekiebox committed:
Remove comments. Updated demo source files.

koekiebox committed:
Merge branch 'main' into issues/68-merge-commit-transfer

koekiebox committed:
Changes include;
* Updated comments
* Default for Transfer fields not required for Commit
* Fix demo_*

koekiebox committed:
Changes include;
* Removed dead code
* Remove references to commit
* Rollback for posted Transfer

sentientwaffle commented on src/state_machine.zig:
```diff
@@ -500,13 +504,13 @@ pub const StateMachine = struct {
 
 // TODO Optimize this by precomputing hashes outside and before committing to the state machine.
-// If we see that a batch of commits contains commits with preimages, then we will:
+// If we see that a batch of posted contains posted with preimages, then we will:
```
`s/posted/posted transfers` (since "posted" isn't a noun like "commits" is).

koekiebox replied:
:100: :bowing_man:

sentientwaffle commented on src/state_machine.zig:
```diff
@@ -337,151 +331,161 @@ pub const StateMachine = struct {
+                return .already_committed;
+            }
 
-        // TODO We need a lookup before inserting in case transfer exists and would overflow limits.
```
This should be implemented (or at least leave the TODO in place). Specifically, before incrementing `{credits,debits}_posted` it needs to verify that adding `{credits,debits}_pending + transfer.amount` wouldn't overflow the integer.

And incrementing `{credits,debits}_pending` should be checked as well.

koekiebox replied:
@sentientwaffle The logic has been updated to perform the `Transfer` lookup/validation prior to exceeding limit checks.
Once the `exists` check is performed, the `{debits/credits}_exceed_{credits/debits}` is performed. 
The `{debits/credits}_exceed_{credits/debits}` verification performs a validation on the sum of the `posted` and `pending` against the account `credits_posted`/`debits_posted`. 
> So no need to check `{credits,debits}_pending`? **Unless I am missing something?**

```zig
// 1. Ensure the transfer is not duplicated prior to limit verification's.
const insert = self.transfers.getOrPutAssumeCapacity(t.id);
            if (insert.found_existing) {
                const exists = insert.value_ptr.*;
                if (exists.debit_account_id != t.debit_account_id) {
                    return .exists_with_different_debit_account_id;
                } else if (exists.credit_account_id != t.credit_account_id) {
                    return .exists_with_different_credit_account_id;
                }
                if (exists.amount != t.amount) return .exists_with_different_amount;
                if (@bitCast(u32, exists.flags) != @bitCast(u32, t.flags)) {
                    return .exists_with_different_flags;
                }
                if (exists.user_data != t.user_data) return .exists_with_different_user_data;
                if (!equal_32_bytes(exists.reserved, t.reserved)) {
                    return .exists_with_different_reserved_field;
                }
                if (exists.timeout != t.timeout) return .exists_with_different_timeout;
                return .exists;
            }

// 2. Verify limits, if limits are exceeded, the transfer is removed. 
            if (dr.debits_exceed_credits(t.amount)) {
                assert(self.transfers.remove(t.id));
                return .exceeds_credits;
            } else if (cr.credits_exceed_debits(t.amount)) {
                assert(self.transfers.remove(t.id));
                return .exceeds_debits;
            }

// 3. Lastly we increment the amounts
insert.value_ptr.* = t;
            if (t.flags.pending) {
                dr.debits_pending += t.amount;
                cr.credits_pending += t.amount;
            } else {
                dr.debits_posted += t.amount;
                cr.credits_posted += t.amount;
            }
            self.commit_timestamp = t.timestamp;
            return .ok;
```

sentientwaffle replied:
Maybe I misunderstood what this `TODO` was referring to. But what I mean is that we need to verify that `self.debits_pending + self.debits_posted + amount` (and likewise for `credits`) doesn't overflow a `u64`, since that would panic.

koekiebox replied:
@sentientwaffle You are right! :bowing_man:  I have updated the `tigerbeetle.zig` to handle the overflow. I have also added a test case for it `test "exceed credit/debit limits panic"`.

koekiebox committed:
Changes include;
* Update DESIGN.md

ifreund commented on src/state_machine.zig:
```diff
@@ -337,151 +331,161 @@ pub const StateMachine = struct {
+            if (t.flags.padding != 0) return .reserved_flag_padding;
 
-        if (t.debit_account_id == t.credit_account_id) return .accounts_are_the_same;
+            var lookup = self.get_transfer(t.id) orelse return .transfer_not_found;
```
```suggestion
            const lookup = self.get_transfer(t.id) orelse return .transfer_not_found;
```

ifreund commented on src/state_machine.zig:
```diff
@@ -337,151 +331,161 @@ pub const StateMachine = struct {
+            assert(!cr.credits_exceed_debits(0));
+
+            // TODO We can combine this lookup with the previous lookup if we return `error!void`:
+            var insert = self.posted.getOrPutAssumeCapacity(t.id);
```
```suggestion
            const insert = self.posted.getOrPutAssumeCapacity(t.id);
```

ifreund commented on src/state_machine.zig:
```diff
@@ -337,151 +331,161 @@ pub const StateMachine = struct {
-                return .exists_with_different_reserved_field;
+
+            var dr = self.get_account(lookup.debit_account_id) orelse return .debit_account_not_found;
+            var cr = self.get_account(lookup.credit_account_id) orelse return .credit_account_not_found;
```
```suggestion
            const dr = self.get_account(lookup.debit_account_id) orelse return .debit_account_not_found;
            const cr = self.get_account(lookup.credit_account_id) orelse return .credit_account_not_found;
```

ifreund commented on src/state_machine.zig:
```diff
@@ -337,151 +331,161 @@ pub const StateMachine = struct {
+            // 3. relating to debtor and creditor.
+            // We use them to distinguish between `cr` (credit account), and `c` (commit).
+            var dr = self.get_account(t.debit_account_id) orelse return .debit_account_not_found;
+            var cr = self.get_account(t.credit_account_id) orelse return .credit_account_not_found;
```
```suggestion
            const dr = self.get_account(t.debit_account_id) orelse return .debit_account_not_found;
            const cr = self.get_account(t.credit_account_id) orelse return .credit_account_not_found;
```

ifreund commented on src/state_machine.zig:
```diff
@@ -337,151 +331,161 @@ pub const StateMachine = struct {
+
+            if (dr.unit != cr.unit) return .accounts_have_different_units;
+
+            var insert = self.transfers.getOrPutAssumeCapacity(t.id);
```
```suggestion
            const insert = self.transfers.getOrPutAssumeCapacity(t.id);
```

ifreund commented on src/state_machine.zig:
```diff
@@ -337,151 +331,161 @@ pub const StateMachine = struct {
+            // 1. derived from the Latin past participles of debitum and creditum, debere and credere,
+            // 2. standing for debit record and credit record, or
+            // 3. relating to debtor and creditor.
+            // We use them to distinguish between `cr` (credit account), and `c` (commit).
```
Since commits don't exist anymore and there's no longer a conflict with `c`, should we switch to just `d`/`c` here and drop the comment?

jorangreef replied:
No, `dr` and `cr` are domain-specific shorthand for debit and credit account respectively (it's literally how the accounts in a double-entry journal entry would be designated if written on paper), so they should stay.

We can drop this part of the comment though: `// We use them to distinguish between "cr" (credit account), and "c" (commit).` since that's no longer relevant.

ifreund commented on src/state_machine.zig:
```diff
@@ -337,151 +331,161 @@ pub const StateMachine = struct {
+            var insert = self.posted.getOrPutAssumeCapacity(t.id);
+            if (insert.found_existing) {
+                unreachable;
+            } else {
```
this could be written as an assertion instead of an if statement, which would save some indentation
```suggestion
            assert(!insert.found_existing);
```

ifreund commented on src/state_machine.zig:
```diff
@@ -337,151 +331,161 @@ pub const StateMachine = struct {
-    }
-
-    fn commit_transfer_rollback(self: *StateMachine, c: Commit) void {
+    fn posted_transfer_rollback(self: *StateMachine, c: Transfer) void {
```
Hmm, I see we actually do still use the `c` variable name here even though the type is now `Transfer`. I wonder if that should be changed so we can switch from `dr`/`cr` to `d`/`c` or if it's clearer like this.

koekiebox replied:
Yeah, I changed my mind since reading your comment. I have now changed it to `ptr` for Posted Transfer Rollback. 
Let me know what you think @ifreund ?

ifreund replied:
I actually find that a bit more confusing as `ptr` is a common abbreviation of "pointer" in zig codebases. Perhaps `posted_transfer` or `pt` would be better?

I also notice that `t`, `dr`, and `cr` should all be made `const` in the function.

koekiebox replied:
Ok, awesome. Thank you @ifreund Updated.

ifreund reviewed (commented):
Here's some largely cosmetic review.

I did notice a fair number of variables that can and should be `const` instead of `var`, though these seem to predate your changes.

jorangreef commented on src/state_machine.zig:
```diff
@@ -127,19 +118,19 @@ pub const StateMachine = struct {
             @intCast(u64, realtime),
         );
-        assert(self.prepare_timestamp > self.commit_timestamp);
+        assert(self.pending_timestamp > self.post_timestamp);
```
These can actually stay as `prepare_timestamp` and `commit_timestamp` because that's what they are—they come from the VSR protocol and relate to VSR prepares and VSR commits respectively.

jorangreef replied:
i.e. They don't relate to two-phase commit transfers. This is also part of the motivation for this PR, so that we don't overload these consensus terms for two-phase payments.

koekiebox replied:
Yeah, I wondered about that. Thank you for the clarification. :-)

jorangreef commented on docs/DESIGN.md:
```diff
@@ -135,9 +135,9 @@ Events are **immutable data structures** that **instantiate or mutate state data
-* An `accepted` amount refers to an amount posted by a committed transfer.
-* A `reserved` amount refers to an inflight amount posted by a two-phace commit transfer only, where the commit is still outstanding, and where the transfer timeout has not yet fired. In other words, the transfer amount has been reserved in the account balance (to avoid double-spending) but not yet committed. The reserved amount will rollback if the transfer ultimately fails.
-* The total debit balance of an account is given by adding `debits_accepted` plus `debits_reserved`. Likewise for the total credit balance of an account.
+* An `posted` amount refers to an amount posted by a committed transfer.
```
```suggestion
* A `posted` amount refers to an amount posted by a transfer.
```

jorangreef commented on docs/DESIGN.md:
```diff
@@ -135,9 +135,9 @@ Events are **immutable data structures** that **instantiate or mutate state data
-* A `reserved` amount refers to an inflight amount posted by a two-phace commit transfer only, where the commit is still outstanding, and where the transfer timeout has not yet fired. In other words, the transfer amount has been reserved in the account balance (to avoid double-spending) but not yet committed. The reserved amount will rollback if the transfer ultimately fails.
-* The total debit balance of an account is given by adding `debits_accepted` plus `debits_reserved`. Likewise for the total credit balance of an account.
+* An `posted` amount refers to an amount posted by a committed transfer.
+* A `pending` amount refers to an inflight amount posted by a two-phace commit transfer only, where the commit is still outstanding, and where the transfer timeout has not yet fired. In other words, the transfer amount has been reserved in the account balance (to avoid double-spending) but not yet committed. The reserved amount will rollback if the transfer ultimately fails.
```
```suggestion
* A `pending` amount refers to an inflight amount yet-to-be-posted by a two-phase commit transfer only, where the commit is still outstanding, and where the transfer timeout has not yet fired. In other words, the transfer amount has been reserved in the account balance (to avoid double-spending) but not yet committed. The reserved amount will rollback if the transfer ultimately fails. By default, transfers post automatically, but being able to reserve the amount as pending and then post the amount only later can sometimes be convenient, for example when switching credit card payments.
```

jorangreef commented on docs/DESIGN.md:
```diff
@@ -135,9 +135,9 @@ Events are **immutable data structures** that **instantiate or mutate state data
-* The total debit balance of an account is given by adding `debits_accepted` plus `debits_reserved`. Likewise for the total credit balance of an account.
+* An `posted` amount refers to an amount posted by a committed transfer.
+* A `pending` amount refers to an inflight amount posted by a two-phace commit transfer only, where the commit is still outstanding, and where the transfer timeout has not yet fired. In other words, the transfer amount has been reserved in the account balance (to avoid double-spending) but not yet committed. The reserved amount will rollback if the transfer ultimately fails.
+* The total debit balance of an account is given by adding `debits_posted` plus `debits_pending`. Likewise for the total credit balance of an account.
```
```suggestion
* The debit balance of an account is given by adding `debits_posted` plus `debits_pending`. Likewise for the credit balance of an account.
```

jorangreef commented on src/benchmark.zig:
```diff
@@ -101,14 +100,14 @@ pub fn main() !void {
             .reserved = [_]u8{0} ** 32,
             .code = 0,
-            .flags = if (IS_TWO_PHASE_COMMIT) .{ .two_phase_commit = true } else .{},
+            .flags = if (IS_TWO_PHASE_COMMIT) .{ .posting = true } else .{},
```
If it's a two-phase payment, then the flag should rather be `.pending = true`, since one-phase payments are posting by default.

jorangreef commented on src/demo_04_create_transfers_two_phase_commit.zig:
```diff
@@ -16,7 +16,7 @@ pub fn main() !void {
             .code = 0,
             .flags = .{
-                .two_phase_commit = true,
+                .pending = true,// Setting Transfre to be 2-phase.
```
```suggestion
                .pending = true, // The transfer is going to be a two-phase transfer.
```

jorangreef commented on src/demo_04_create_transfers_two_phase_commit.zig:
```diff
@@ -29,7 +29,7 @@ pub fn main() !void {
             .code = 0,
             .flags = .{
-                .two_phase_commit = true,
+                .pending = true,// Setting Transfre to be 2-phase.
```
```suggestion
                .pending = true,
```

jorangreef commented on src/demo_04_create_transfers_two_phase_commit.zig:
```diff
@@ -43,7 +43,7 @@ pub fn main() !void {
             .code = 0,
             .flags = .{
-                .two_phase_commit = true,
+                .pending = true,// Setting Transfre to be 2-phase.
```
```suggestion
                .pending = true,
```

jorangreef commented on src/demo_05_accept_transfers.zig:
```diff
@@ -1,23 +1,23 @@
             .reserved = [_]u8{0} ** 32,
             .code = 0,
-            .flags = .{},
+            .flags = .{ .post_pending_transfer = true },// Post the pending 2-phase Transfer.
```
```suggestion
            .flags = .{ .post_pending_transfer = true }, // Post the pending two-phase transfer.
```

jorangreef commented on src/demo_05_accept_transfers.zig:
```diff
@@ -1,23 +1,23 @@
             .reserved = [_]u8{0} ** 32,
             .code = 0,
-            .flags = .{},
+            .flags = .{ .post_pending_transfer = true },// Post the pending 2-phase Transfer.
```
```suggestion
            .flags = .{ .post_pending_transfer = true },
```

We can just comment the first time on line 12 as we already do, when we set this flag for the first time. Thereafter, we don't need a comment.

jorangreef replied:
Also, since there's not too much vertical space between the use of these flags.

jorangreef commented on src/demo_05_accept_transfers.zig:
```diff
@@ -1,23 +1,23 @@
 const tb = @import("tigerbeetle.zig");
```
Let's rename these demo file names:

`demo_04_create_pending_transfers.zig`
`demo_05_post_pending_transfers.zig`
`demo_06_void_pending_transfers.zig`

And also update any link references in the docs, there may be several links to them from across the repo.

jorangreef commented on src/tigerbeetle.zig:
```diff
@@ -58,17 +58,17 @@ pub const AccountFlags = packed struct {
     code: u32,
     flags: TransferFlags,
-    amount: u64,
+    amount: u64 = 0,
```
We shouldn't default these specific values in this case.

Otherwise, it's too easy for the developer to forget to initialize them. For example, they might forget to set the amount and this would make it possible for a one-phase transfer to accidentally get posted with a zero amount instead of the real amount.

Another reason is that for one-phase transfers, zero is not a valid value for `debit/credit_account_id`.

jorangreef replied:
For these structs, the developer should always have to think explicitly about each field and initialize them clearly.

koekiebox replied:
Ok understood. Agreed. I thought to make the `Transfer` commit more lenient, but the result would be a nasty 1-phase `Transfer`.

jorangreef commented on src/tigerbeetle.zig:
```diff
@@ -78,40 +78,17 @@ pub const Transfer = packed struct {
+    post_pending_transfer: bool = false,
+    void_pending_transfer: bool = false,
+    hashlock: bool = false,
+    padding: u27 = 0,
```
Nice!

jorangreef commented on src/tigerbeetle.zig:
```diff
@@ -152,28 +129,19 @@ pub const CreateTransferResult = enum(u32) {
+    // Fields for the 2-phase Transfer
+    cannot_void_and_post_two_phase_commit,
     transfer_not_found,
     transfer_not_two_phase_commit,
```
```suggestion
    transfer_not_pending,
```

jorangreef commented on src/tigerbeetle.zig:
```diff
@@ -152,28 +129,19 @@ pub const CreateTransferResult = enum(u32) {
-    linked_event_failed,
-    reserved_field,
-    reserved_flag_padding,
+    // Fields for the 2-phase Transfer
```
```suggestion
    // For two-phase transfers:
```

jorangreef commented on src/tigerbeetle.zig:
```diff
@@ -152,28 +129,19 @@ pub const CreateTransferResult = enum(u32) {
-    credit_amount_was_not_reserved,
-    exceeds_credits,
-    exceeds_debits,
+    already_committed,
```
```suggestion
    transfer_already_posted,
```

jorangreef commented on src/tigerbeetle.zig:
```diff
@@ -152,28 +129,19 @@ pub const CreateTransferResult = enum(u32) {
     transfer_not_two_phase_commit,
-    transfer_expired,
-    already_committed,
     already_committed_but_accepted,
```
```suggestion
    transfer_already_posted,
```

jorangreef commented on src/tigerbeetle.zig:
```diff
@@ -152,28 +129,19 @@ pub const CreateTransferResult = enum(u32) {
-    transfer_expired,
-    already_committed,
     already_committed_but_accepted,
     already_committed_but_rejected,
```
```suggestion
    transfer_already_voided,
```

jorangreef commented on src/tigerbeetle.zig:
```diff
@@ -152,28 +129,19 @@ pub const CreateTransferResult = enum(u32) {
-    preimage_requires_condition,
     preimage_invalid,
+    preimage_requires_condition,
+    debit_amount_was_not_reserved,
```
```suggestion
    debit_amount_not_pending,
```

jorangreef commented on src/tigerbeetle.zig:
```diff
@@ -152,28 +129,19 @@ pub const CreateTransferResult = enum(u32) {
     preimage_invalid,
+    preimage_requires_condition,
+    debit_amount_was_not_reserved,
+    credit_amount_was_not_reserved,
```
```suggestion
    credit_amount_not_pending,
```

jorangreef commented on src/tigerbeetle.zig:
```diff
@@ -152,28 +129,19 @@ pub const CreateTransferResult = enum(u32) {
-    reserved_field,
-    reserved_flag_padding,
+    // Fields for the 2-phase Transfer
+    cannot_void_and_post_two_phase_commit,
```
```suggestion
    cannot_post_and_void_pending_transfer,
```

jorangreef commented on src/tigerbeetle.zig:
```diff
@@ -152,28 +129,19 @@ pub const CreateTransferResult = enum(u32) {
     exceeds_debits,
     two_phase_commit_must_timeout,
```
```suggestion
    pending_transfer_must_timeout,
```

jorangreef replied:
We should also rename these:

`exists_and_already_committed_and_accepted`
`exists_and_already_committed_and_rejected`

To:

`exists_and_already_posted`
`exists_and_already_voided`

jorangreef commented on src/tigerbeetle.zig:
```diff
@@ -152,28 +129,19 @@ pub const CreateTransferResult = enum(u32) {
     exceeds_debits,
     two_phase_commit_must_timeout,
     timeout_reserved_for_two_phase_commit,
```
```suggestion
    timeout_reserved_for_pending_transfer,
```

jorangreef commented on src/state_machine.zig:
```diff
@@ -36,17 +31,16 @@ pub const StateMachine = struct {
-    prepare_timestamp: u64,
-    commit_timestamp: u64,
+    pending_timestamp: u64,
+    post_timestamp: u64,
```
These should actually keep their prepare/commit names, since they are used by VSR in that context.

koekiebox replied:
:heavy_check_mark: :100:

jorangreef commented on src/state_machine.zig:
```diff
@@ -62,34 +56,33 @@ pub const StateMachine = struct {
-        try commits.ensureTotalCapacity(@intCast(u32, commits_max));
+        var posted = HashMapPosted.init(allocator);
+        errdefer posted.deinit();
+        try posted.ensureTotalCapacity(@intCast(u32, commits_max));
```
```suggestion
        try posted.ensureTotalCapacity(@intCast(u32, pending_max));
```

koekiebox replied:
:100: :+1:

jorangreef reviewed (changes requested):
Nice work @koekiebox !

I've done a first round of review—we can tweak some of our state machine result codes to also get rid of the `commit` terminology.

koekiebox committed:
Apply suggestions from code review

Applied suggested review comments.

Co-authored-by: Joran Dirk Greef <joran@coil.com>
Co-authored-by: Isaac Freund <isaac.freund@coil.com>

koekiebox committed:
Review comments and document updates.

koekiebox committed:
Revert to prepare_timestamp and commit_timestamp due to the timestamp relating to the VSR protocol.

koekiebox committed:
Review comments from Joran.

koekiebox committed:
Review comments from Isaac and DJ.

koekiebox committed:
Changes include:
* Rename ptr to pt
* Make `var` a `const` where applicable

koekiebox committed:
Changes include:
* panic for overflow picked up by DJ

jorangreef commented on docs/DEEP_DIVE.md:
````diff
@@ -97,14 +97,14 @@ Let's commit (and accept):
+zig/zig run src/demo_06_void_pending_transfers.zig
 ```
 
 **From here, feel free to tweak these demos and see what happens. You can explore all our accounting invariants (and the DSL we created for these) in `src/state_machine.zig` by grepping the source for `fn create_account`, `fn create_transfer`, and `fn commit_transfer`.**
````
```suggestion
**From here, feel free to tweak these demos and see what happens. You can explore all our accounting invariants (and the DSL we created for these) in `src/state_machine.zig` by grepping the source for `fn create_account` or `fn create_transfer`.**
```

jorangreef commented on docs/DEEP_DIVE.md:
````diff
@@ -97,14 +97,14 @@ Let's commit (and accept):
 zig/zig run src/demo_02_lookup_accounts.zig
 ```
 
 Let's also pretend someone else tried to commit (but reject) concurrently:
````
```suggestion
Let's also pretend someone else tried to void the pending transfer concurrently:
```

There may be a few more places for the text under the `Demo 4, 5, 6: Two-phase commit journal entries` header, which could also become `Demo 4, 5, 6: Two-phase journal entries`

jorangreef commented on docs/DESIGN.md:
```diff
@@ -135,9 +135,9 @@ Events are **immutable data structures** that **instantiate or mutate state data
-* A `reserved` amount refers to an inflight amount posted by a two-phace commit transfer only, where the commit is still outstanding, and where the transfer timeout has not yet fired. In other words, the transfer amount has been reserved in the account balance (to avoid double-spending) but not yet committed. The reserved amount will rollback if the transfer ultimately fails.
-* The total debit balance of an account is given by adding `debits_accepted` plus `debits_reserved`. Likewise for the total credit balance of an account.
+* A `posted` amount refers to an amount posted by a transfer.
+* A `pending` amount refers to an inflight amount yet-to-be-posted by a two-phase commit transfer only, where the commit is still outstanding, and where the transfer timeout has not yet fired. In other words, the transfer amount has been reserved in the account balance (to avoid double-spending) but not yet committed. The reserved amount will rollback if the transfer ultimately fails. By default, ...
```
```suggestion
* A `pending` amount refers to an inflight amount yet-to-be-posted by a two-phase transfer only, where the transfer is still pending, and where the transfer timeout has not yet fired. In other words, the transfer amount has been reserved in the pending account balance (to avoid double-spending) but not yet posted to the posted balance. The reserved amount will rollback if the transfer ultimately fails. By default, transfers post automatically, but being able to reserve the amount as pending and then post the amount only later can sometimes be convenient, for example when switching credit card payments.
```

jorangreef commented on src/config.zig:
```diff
@@ -53,9 +53,9 @@ pub const transfers_max = switch (deployment_environment) {
 };
 
-/// The maximum number of two-phase commits to store in memory:
+/// The maximum number of two-phase Transfers to store in memory:
```
```suggestion
/// The maximum number of two-phase transfers to store in memory:
```

jorangreef commented on src/config.zig:
```diff
@@ -53,9 +53,9 @@ pub const transfers_max = switch (deployment_environment) {
+/// The maximum number of two-phase Transfers to store in memory:
 /// This impacts the amount of memory allocated at initialization by the server.
-pub const commits_max = transfers_max;
+pub const pending_max = transfers_max;
```
```suggestion
pub const transfers_pending_max = transfers_max;
```

This is one of the first config variables we had :)

This limit will eventually fall away, as the LSM engine comes in, allowing for data to spill to disk.

jorangreef commented on src/demo_04_create_pending_transfers.zig:
```diff
@@ -16,7 +16,7 @@ pub fn main() !void {
             .code = 0,
             .flags = .{
-                .two_phase_commit = true,
+                .pending = true, // Setting Transfer to be 2-phase.
```
```suggestion
                .pending = true, // Set this transfer to be two-phase.
```

jorangreef commented on src/demo_04_create_pending_transfers.zig:
```diff
@@ -29,7 +29,7 @@ pub fn main() !void {
             .code = 0,
             .flags = .{
-                .two_phase_commit = true,
+                .pending = true, // Setting Transfer to be 2-phase.
```
```suggestion
                .pending = true, // Set this transfer to be two-phase.
```

Since it's consistent with the "Link this transfer" formula and casing on line 33.

jorangreef commented on src/demo_04_create_pending_transfers.zig:
```diff
@@ -43,7 +43,7 @@ pub fn main() !void {
             .code = 0,
             .flags = .{
-                .two_phase_commit = true,
+                .pending = true, // Setting Transfer to be 2-phase.
```
```suggestion
                .pending = true, // Set this transfer to be two-phase.
```

jorangreef commented on src/demo_05_post_pending_transfers.zig:
```diff
@@ -0,0 +1,33 @@
+            .reserved = [_]u8{0} ** 32,
+            .timeout = 0,
+            .code = 0,
+            .flags = .{ .post_pending_transfer = true }, // Post the pending 2-phase Transfer.
```
```suggestion
            .flags = .{ .post_pending_transfer = true }, // Post the pending two-phase transfer.
```

jorangreef commented on src/demo_05_post_pending_transfers.zig:
```diff
@@ -0,0 +1,33 @@
+            .timeout = 0,
+            .code = 0,
+            .flags = .{ .post_pending_transfer = true }, // Post the pending 2-phase Transfer.
+            .amount = 0, // Amount from pending Transfer.
```
```suggestion
            .amount = 0, // Inherit the amount from the pending transfer.
```

jorangreef commented on src/demo_05_post_pending_transfers.zig:
```diff
@@ -0,0 +1,33 @@
+            .reserved = [_]u8{0} ** 32,
+            .timeout = 0,
+            .code = 0,
+            .flags = .{ .post_pending_transfer = true }, // Post the pending 2-phase Transfer.
```
```suggestion
            .flags = .{ .post_pending_transfer = true }, // Post the pending two-phase transfer.
```

jorangreef commented on src/demo_05_post_pending_transfers.zig:
```diff
@@ -0,0 +1,33 @@
+            .timeout = 0,
+            .code = 0,
+            .flags = .{ .post_pending_transfer = true }, // Post the pending 2-phase Transfer.
+            .amount = 0, // Amount from pending Transfer.
```
```suggestion
            .amount = 0, // Inherit the amount from the pending transfer.
```

jorangreef commented on src/state_machine.zig:
```diff
@@ -260,8 +249,13 @@ pub const StateMachine = struct {
                 .create_accounts => self.create_account_rollback(event),
-                .create_transfers => self.create_transfer_rollback(event),
-                .commit_transfers => self.commit_transfer_rollback(event),
+                .create_transfers => {
```
Let's rather switch on `event.flags.pending` internally within the `create_transfer_rollback` handler, since that's where we handle operation-specific decisions. That will then make this `switch(operation)` branch symmetrical with the original corresponding `switch(operation)` branch that it rolls back.

jorangreef replied:
It will also help to keep `rollback()` itself simpler, which is good as it already deals with the tricky chaining logic.

jorangreef replied:
This should only call out to a single `_rollback` method here, as described above.

koekiebox replied:
Fixed.

jorangreef commented on src/state_machine.zig:
```diff
@@ -339,62 +333,124 @@ pub const StateMachine = struct {
-            return .timeout_reserved_for_two_phase_commit;
-        }
-        if (!t.flags.condition and !zeroed_32_bytes(t.reserved)) return .reserved_field;
+        // Either a 2-phase transfer post/void
```
```suggestion
```

jorangreef replied:
The branch conditions and error code are clear.

jorangreef commented on src/state_machine.zig:
```diff
@@ -339,62 +333,124 @@ pub const StateMachine = struct {
-        }
-        if (!t.flags.condition and !zeroed_32_bytes(t.reserved)) return .reserved_field;
+        // Either a 2-phase transfer post/void
+        if (t.flags.post_pending_transfer and t.flags.void_pending_transfer) {
```
This branch should move within the `t.flags.post_pending_transfer or t.flags.void_pending_transfer` branch since it relates there, and also to improve performance for the common case of one-phase transfers.

jorangreef replied:
We can then also refactor the two-phase logic out into `create_transfer_two_phase` where we handle the posting/voiding of the pending transfer.

koekiebox replied:
Fixed.

jorangreef commented on src/state_machine.zig:
```diff
@@ -339,62 +333,124 @@ pub const StateMachine = struct {
+            if (t.flags.padding != 0) return .reserved_flag_padding;
 
-        if (t.amount == 0) return .amount_is_zero;
+            const lookup = self.get_transfer(t.id) orelse return .transfer_not_found;
```
```suggestion
            const pending = self.get_transfer(t.user_data) orelse return .transfer_not_found;
```

The second posting transfer's id should actually be its own unique id, not the id of the pending transfer.

The id for the first pending transfer should then instead be stored in `user_data`.

jorangreef replied:
We then also need to do an idempotency check on the second posting transfer's id.

jorangreef replied:
The motivation for using `user_data` in this way, is so that all pending/posted transfers are then immutable, with their own IDs and with their own timestamps.

koekiebox replied:
Resolved.

jorangreef commented on src/state_machine.zig:
```diff
@@ -339,62 +333,124 @@ pub const StateMachine = struct {
+
+            if (t.debit_account_id == t.credit_account_id) return .accounts_are_the_same;
+
+            // The etymology of the DR and CR abbreviations for debit and credit is interesting, either:
```
```suggestion
            // The etymology of the DR and CR abbreviations for debit/credit is interesting, either:
```

Column length fix.

jorangreef commented on src/state_machine.zig:
```diff
@@ -339,62 +333,124 @@ pub const StateMachine = struct {
+            if (t.debit_account_id == t.credit_account_id) return .accounts_are_the_same;
+
+            // The etymology of the DR and CR abbreviations for debit and credit is interesting, either:
+            // 1. derived from the Latin past participles of debitum and creditum, debere and credere,
```
```suggestion
            // 1. derived from the Latin past participles of debitum/creditum, i.e. debere/credere,
```

jorangreef commented on src/state_machine.zig:
```diff
@@ -339,62 +333,124 @@ pub const StateMachine = struct {
+            if (self.get_commit(t.id)) |exists| {
+                if (!exists.flags.void_pending_transfer and t.flags.void_pending_transfer) return .transfer_already_posted;
+                if (exists.flags.void_pending_transfer and !t.flags.void_pending_transfer) return .transfer_already_voided;
+                return .transfer_already_posted;
```
```suggestion
                if (exists.flags.post_pending_transfer) return .transfer_already_posted;
                if (exists.flags.void_pending_transfer) return .transfer_already_voided;
                return .exists_with_different_flags;
```

koekiebox replied:
Resolved.

sentientwaffle commented on src/state_machine.zig:
```diff
@@ -1458,6 +1458,37 @@ test "create/lookup/rollback 2-phase transfers" {
     try testing.expectEqual(@as(u64, 75), account_2_rollback_reject.credits_pending);
 }
 
+test "exceed credit/debit limits panic" {
```
```suggestion
test "credit/debit limit overflows " {
```

sentientwaffle commented on src/tigerbeetle.zig:
```diff
@@ -26,12 +26,12 @@ pub const Account = packed struct {
     pub fn credits_exceed_debits(self: *const Account, amount: u64) bool {
         return (self.flags.credits_must_not_exceed_debits and
-            self.credits_pending + self.credits_posted + amount > self.debits_posted);
+            (@as(u66, self.credits_pending) + @as(u66, self.credits_posted) + @as(u66, amount)) > self.debits_posted);
```
I wonder if the overflow checks should be separate from (and prior to) the "exceeds credits/debits" checks, so that they can have a separate error (e.g. `overflows_{credits,debits}_{posted,pending}`). @jorangreef what do you think?

jorangreef replied:
Agreed @sentientwaffle — it's good to have a separate error and to handle the overflow before checking the limit. Semantically, they're different errors, and the operator may want to handle them differently.

koekiebox replied:
Resolved.

sentientwaffle commented on src/state_machine.zig:
```diff
@@ -1458,6 +1458,37 @@ test "create/lookup/rollback 2-phase transfers" {
+        .debits_posted = std.math.maxInt(u64),
+        .credits_pending = std.math.maxInt(u64),
+        .credits_posted = std.math.maxInt(u64),
+    };
```
Can you make these tests nearer to the actual boundary? e.g. test that it overflows at `maxInt(u64)+1` but _doesn't_ overflow at `maxInt(u64)`. And verify that the combination of `pending+posted+amount` triggers it (e.g. test 0 pending + high posted + 1, and also high pending + 0 posted + 1).

koekiebox replied:
Resolved.

jorangreef commented on src/state_machine.zig:
```diff
@@ -339,62 +333,124 @@ pub const StateMachine = struct {
+                return .transfer_already_posted;
+            }
 
-        if (dr.unit != cr.unit) return .accounts_have_different_units;
```
We should still rename `unit` to `ledger` and add `ledger` to Transfer as part of this PR.

Then here, if the `ledger` value is non-zero, we must check that it is the same as that of the pending transfer. Same for `code`.

koekiebox replied:
Resolved.

jorangreef reviewed (commented):
Thanks @koekiebox here's the second round!

koekiebox committed:
Apply suggestions from code review

Co-authored-by: Joran Dirk Greef <joran@coil.com>
Co-authored-by: djg <sentientwaffle@gmail.com>

koekiebox committed:
Issue #68:
* Update docs
* Introduce HashTable for transaction via user_data lookup for 2 phase transfer
* Review feedback changes

koekiebox committed:
Updates include:
* Make use of user_data for 2-phase transfer.
* Fixed README.md wording.
* Introduce limit check result enum.

koekiebox committed:
Changes include:
* Limit check to use enum for result.

koekiebox committed:
Updates include:
* Make use of new Transfer structure
* Updated test cases (still broken)
* Remove the need for additional HashMap.

sentientwaffle commented on src/tigerbeetle.zig:
```diff
@@ -10,28 +10,36 @@ pub const Account = packed struct {
-    pub fn debits_exceed_credits(self: *const Account, amount: u64) bool {
-        return (self.flags.debits_must_not_exceed_credits and
-            self.debits_reserved + self.debits_accepted + amount > self.credits_accepted);
+    pub fn debits_exceed_credits(self: *const Account, amount: u64) LimitExceedResult {
```
I think the overflow check shouldn't be part of `debits_exceed_credits`, it should be a separate function on `Account`. This will also fix (and remove the possibility of) incorrectly checking `debits_must_not_exceed_credits` prior to the overflow test (as we do now). The overflow check must precede the `debits_exceed_credits` check.

Likewise, an overflow should return its own error code (e.g. `CreateTransferResult.debits_overflow`).

(And likewise for `credits_exceed_debits`).

koekiebox replied:
Resolved.

sentientwaffle commented on src/tigerbeetle.zig:
```diff
@@ -10,28 +10,36 @@ pub const Account = packed struct {
+    pub fn debits_exceed_credits(self: *const Account, amount: u64) LimitExceedResult {
+        if (!self.flags.debits_must_not_exceed_credits) return .ok;
+
+        const limit_amount: u66 = (@as(u66, self.debits_pending) + @as(u66, self.debits_posted) + @as(u66, amount));
```
Rather than using `u66`, I suggest using something like this instead:

```
fn try_add(a: u64, b: u64) ?u64 {
  var c: u64 = undefined;
  return if (@addWithOverflow(u64, a, b, &c)) null else c;
}
```

Then in the call site:

```
const limit_amount = try_add(self.debits_pending + self.debits_posted, amount) orelse return .debits_overflow;
```

koekiebox replied:
Thank you for the suggestion @sentientwaffle . I need to study the build-in functions for Zig closer. :-)

koekiebox committed:
Updates include:
* Updated test cases
* Introduce pending_id
* Create certain post fields from pending transfer

koekiebox committed:
Issue #68 with PR #83:
* Review comments from review round 2 DJ and Joran.

koekiebox committed:
Merge branch 'main' into issues/68-merge-commit-transfer

jorangreef commented on src/tigerbeetle.zig:
```diff
@@ -131,49 +133,42 @@ pub const CreateTransferResult = enum(u32) {
+    preimage_requires_condition,
+    debit_amount_not_pending,
+    credit_amount_not_pending,
+    pending_id_should_be_zero,
```
```suggestion
    pending_id_must_be_zero,
```

jorangreef replied:
Since SHOULD implies that it's desirable but not required.

koekiebox replied:
Fixed. 🙇

jorangreef commented on src/tigerbeetle.zig:
```diff
@@ -131,49 +133,42 @@ pub const CreateTransferResult = enum(u32) {
     credit_account_not_found,
     accounts_are_the_same,
-    accounts_have_different_units,
+    accounts_have_different_ledgers,
```
We also need `transfer_has_different_ledger` where the accounts have the same `ledger` but where the transfer (or posting transfer) has a different `ledger` to the accounts.

koekiebox replied:
`exists_with_different_ledger` is already there. I don't have that on my branch. 

```zig
if (dr.ledger != cr.ledger) return .accounts_have_different_ledgers;
```

jorangreef commented on src/tigerbeetle.zig:
```diff
@@ -131,49 +133,42 @@ pub const CreateTransferResult = enum(u32) {
     ok,
     linked_event_failed,
     exists,
+    exists_user_data,
```
It doesn't look like this error is used anywhere?

koekiebox replied:
True. I used that from before. 🙇 👍 
Removed. :-)

jorangreef commented on src/tigerbeetle.zig:
```diff
@@ -10,28 +10,42 @@ pub const Account = packed struct {
-    pub fn debits_exceed_credits(self: *const Account, amount: u64) bool {
-        return (self.flags.debits_must_not_exceed_credits and
-            self.debits_reserved + self.debits_accepted + amount > self.credits_accepted);
+    pub fn debits_exceed_credits(self: *const Account, amount: u64) LimitExceedResult {
```
Let's rather keep the simpler `bool` return type here—to reduce dimensionality:

1. This function should not worry about overflow at all.
2. Overflow should be handled by a separate function, called directly by the state machine, before the state machine gets to calling this.
3. The state machine should also return separate errors `.debits_pending/posted_would_overflow` or `.credits_pending/posted_would_overflow` etc.

koekiebox replied:
1. Done.
2. Done.
3. Done. I made the error returns for the enum: `overflow_credits/overflow_debits`

jorangreef commented on src/tigerbeetle.zig:
```diff
@@ -10,28 +10,42 @@ pub const Account = packed struct {
-    unit: u16,
     /// A chart of accounts code describing the type of account (e.g. clearing, settlement):
+    ledger: u16,
+    /// The currency code/type related to the transfer:
```
```suggestion
```

jorangreef commented on src/tigerbeetle.zig:
```diff
@@ -10,28 +10,42 @@ pub const Account = packed struct {
     /// Reserved for accounting policy primitives:
     reserved: [48]u8,
-    unit: u16,
     /// A chart of accounts code describing the type of account (e.g. clearing, settlement):
```
This comment actually applies to `code` not `ledger`.

The `ledger` exists for isolation within a cluster. e.g. All accounts within a ledger have the same currency and transfers may only interact with accounts in the same ledger.

jorangreef commented on src/tigerbeetle.zig:
```diff
@@ -10,28 +10,42 @@ pub const Account = packed struct {
     reserved: [48]u8,
-    unit: u16,
     /// A chart of accounts code describing the type of account (e.g. clearing, settlement):
+    ledger: u16,
```
The account `ledger` must also be a `u32` the same as for a transfer.

`AccountFlags` padding must be adjusted so that `AccountFlags` is `sizeOf(u16)`.

koekiebox replied:
Fixed.

jorangreef commented on src/state_machine.zig:
```diff
@@ -62,34 +57,26 @@ pub const StateMachine = struct {
-            .transfers = transfers,
-            .commits = commits,
-        };
+        return StateMachine{ .allocator = allocator, .prepare_timestamp = 0, .commit_timestamp = 0, .accounts = accounts, .transfers = transfers, .posted = posted };
```
```suggestion
        return StateMachine{ .allocator = allocator, .prepare_timestamp = 0, .commit_timestamp = 0, .accounts = accounts, .transfers = transfers, .posted = posted, };
```

Then run `zig fmt` to reflow the line.

We must keep lines to a hard line limit of 100 columns.

koekiebox replied:
Fixed.

jorangreef commented on src/state_machine.zig:
```diff
@@ -309,13 +297,22 @@ pub const StateMachine = struct {
 
         // Opening balances may never exceed limits:
-        if (a.debits_exceed_credits(0)) return .exceeds_credits;
-        if (a.credits_exceed_debits(0)) return .exceeds_debits;
```
Let's keep this.

Rather add the overflow checks above these lines, with separate errors.

jorangreef commented on src/state_machine.zig:
```diff
@@ -339,149 +336,174 @@ pub const StateMachine = struct {
 
-    fn commit_transfer(self: *StateMachine, c: Commit) CommitTransferResult {
-        assert(c.timestamp > self.commit_timestamp);
+            const c_acc_id = if (t.credit_account_id == 0) lookup.credit_account_id else t.credit_account_id;
```
```suggestion
            const cr_id = if (t.credit_account_id == 0) lookup.credit_account_id else t.credit_account_id;
```

koekiebox replied:
Fixed.

jorangreef commented on src/state_machine.zig:
```diff
@@ -339,149 +336,174 @@ pub const StateMachine = struct {
-        }
-        assert(self.transfers.remove(t.id));
-    }
+            const d_acc_id = if (t.debit_account_id == 0) lookup.debit_account_id else t.debit_account_id;
```
```suggestion
            const dr_id = if (t.debit_account_id == 0) lookup.debit_account_id else t.debit_account_id;
```

koekiebox replied:
Fixed.

jorangreef commented on src/state_machine.zig:
```diff
@@ -339,149 +336,174 @@ pub const StateMachine = struct {
-            if (exists.flags.reject and !c.flags.reject) return .already_committed_but_rejected;
-            return .already_committed;
-        }
+            const amnt = if (t.amount == 0) lookup.amount else t.amount;
```
```suggestion
            const amount = if (t.amount == 0) lookup.amount else t.amount;
```

koekiebox replied:
Fixed.

jorangreef commented on src/tigerbeetle.zig:
```diff
@@ -131,49 +133,42 @@ pub const CreateTransferResult = enum(u32) {
     condition_requires_preimage,
-    preimage_requires_condition,
     preimage_invalid,
+    preimage_requires_condition,
```
We should drop any errors relating to `condition` and `preimage` now, since we've dropped hash locks for the time being.

koekiebox replied:
Done.

jorangreef commented on src/tigerbeetle.zig:
```diff
@@ -131,49 +133,42 @@ pub const CreateTransferResult = enum(u32) {
+    credit_amount_not_pending,
+    pending_id_should_be_zero,
+    pending_id_is_zero,
+    post_amount_exceeds_pending_amount,
```
```suggestion
    amount_exceeds_pending_amount,
```

koekiebox replied:
Fixed. 🙇

jorangreef reviewed (commented):
Thanks @koekiebox!

koekiebox committed:
Issue #63:
* 3rd round of review comments.
* Single item remains.

koekiebox committed:
Issue #63:
* Split the overflow and limit checks
* Add test cases for the above
* Remove the limit check result
* Fix the unit to ledger rename

koekiebox committed:
Changes include;
* Review from Joran
* Additional overflow checks
* Updated test cases

koekiebox committed:
Changes include;
* Only store flags on posted
* Additional test cases

koekiebox committed:
Merge branch 'main' into issues/68-merge-commit-transfer

koekiebox committed:
Changes include;
* Cleanup of void/posted
* Updated test cases
* Attend to all TODO's
* Remove unnecasary flag to confirm pending

koekiebox committed:
Changes include;
* Make ledger u32 as with transfer
* fix formatting

koekiebox committed:
Changes include;
* Fix benchmark.zig

koekiebox committed:
Merge branch 'main' into issues/68-merge-commit-transfer

# Conflicts:
#	src/benchmark.zig

koekiebox committed:
Changes include;
* Fix benchmark.zig now that main was updated.

koekiebox committed:
Changes include;
* Fix demo applications for new transfer/commit structure.

jorangreef committed:
StateMachine: Fix and finish business logic

Fix post_or_void_pending_transfer_rollback() which was always using
t.amount instead of (t.amount or p.amount), and which was using
t.debit/credit_account_id (which may be zero) instead of
p.debit/credit_account_id.

Fix 2 credit account overflow checks that were actually checking the
dr account instead of the cr account.

Fix create_transfer() to check flags.padding, reserved, pending_id.
Some of these checks were missing for post_or_void_pending_transfer()
as well.

Fix create_transfer() to check that the ledger matches the accounts'
ledger, the check was missing, so this invariant was not being
enforced.

Fix post_or_void_pending_transfer() business logic order so that
.exists is returned before a .pending_transfer_already_posted/voided.
The transfer_exists() call was previously dead code and unreachable,
since the get_posted() check would always return before then.

Do not overload .exists results when the intention is to compare the
posting/voiding transfer with the pending transfer. Rather, introduce
explicit results e.g. pending_transfer_has_different_credit_account_id.

Simplify overflow checks to use a single private overflow() helper,
without increasing the surface area of the Account struct.

Only store boolean posted/voided state in HashMapPosted, since that
is all we need (saves 1 byte per pending transfer).

Remove exceeds/overflows test. The test did not have 100% coverage, and
missed the overflow check bugs.

Remove unused HashMapUserDataToId type definition.

Rename transfer_rollback() to create_transfer_rollback().

Use mem alias instead of std.mem.

Use const* Account/Transfer argument types.

Do not allow id/debit/credit_account_id/ledger/code to be zero.
Zero is used to indicate that a posting/voiding transfer must inherit
the value from the pending transfer.

Do not allow both debits/credits_must_not_exceed_debits/credits flags
to be set simultaneously (the flags are mutually exclusive).

Do not compare reserved field for existing objects, since it must always
be zero.

Reorder transfer_exists() checks to match the order of fields as they
are defined in the Transfer struct.

Group CreateTransferResult with newlines according to their usage in the
state machine. Order these result enums in exactly the same order that
they are returned from the state machine to make auditing easier.

jorangreef reviewed (changes requested):
Thanks @koekiebox!

I pushed a commit to fix a few bugs in the business logic (some overflow checks using the `dr` account instead of the `cr` account, rollback using the wrong amount, several checks not being enforced).

The state machine code is now done.

TODO:

1. From here, please would you update the tests to have 100% coverage. **Changing any of the logic by hand in the state machine, with respect to any `CreateTransferResult` should fail the tests.** All results should have 100% coverage. Previously, the tests were not checking all the invariants. For example, it was possible to post a transfer with a different `ledger` to the accounts, and rollback was adding back the wrong amount.

2. You will see that the `CreateTransferResult` enums are also now all carefully grouped and ordered. **Please would you make the tests follow this same order**, so that we can more easily double-check in future that all results are covered.

sentientwaffle commented on src/tigerbeetle.zig:
```diff
@@ -78,102 +81,105 @@ pub const Transfer = packed struct {
+    accounts_must_have_the_same_ledger,
+    transfer_must_have_the_same_ledger_as_accounts,
+
+    overflows_debits_pending,
+    overflows_credits_pending,
+    overflows_debits_posted,
+    overflows_credits_posted,
+    overflows_debits,
+    overflows_credits,
```
nit: These might read easier if reordered:
```suggestion
    overflows_debits,
    overflows_debits_pending,
    overflows_debits_posted,
    overflows_credits,
    overflows_credits_pending,
    overflows_credits_posted,
```

koekiebox replied:
Fixed.

jorangreef replied:
This is a great suggestion and what we would usually do, taking advantage of the way we put the most significant word first so everything lines up.

However, in this case, we should leave the order as is, because it reflects the order that it needs to be checked in the state machine. The reason for this order is that we want to surface the overflow errors at the finest granularity first, i.e.  `overflows_debits_pending` and `overflows_credits_pending` before the others etc.

sentientwaffle commented on src/state_machine.zig:
```diff
@@ -302,186 +291,263 @@ pub const StateMachine = struct {
-        if (!c.flags.reject) {
-            dr.debits_accepted -= t.amount;
-            cr.credits_accepted -= t.amount;
+    fn transfer_exists(t: *const Transfer, e: *const Transfer) CreateTransferResult {
```
Can we assert that the 2 transfers' `id`s match?

koekiebox replied:
Fixed.

sentientwaffle commented on src/state_machine.zig:
```diff
@@ -499,21 +565,29 @@ pub const StateMachine = struct {
-    fn get_commit(self: *StateMachine, id: u128) ?*Commit {
-        return self.commits.getPtr(id);
+    /// Returns whether a pending transfer, if it exists, has already been posted or voided.
+    fn get_posted(self: *StateMachine, pending_id: u128) ?bool {
```
```suggestion
    fn get_posted(self: *const StateMachine, pending_id: u128) ?bool {
```

koekiebox replied:
Fixed.

sentientwaffle commented on src/state_machine.zig:
```diff
@@ -499,21 +565,29 @@ pub const StateMachine = struct {
-    var target: [32]u8 = undefined;
-    std.crypto.hash.sha2.Sha256.hash(&preimage, &target, .{});
-    return std.crypto.utils.timingSafeEql([32]u8, target, condition);
+fn overflow(a: u64, b: u64) bool {
```
Maybe rename to `sum_overflows` or `sum_will_overflow` to be a little more descriptive.

koekiebox replied:
Fixed.

jorangreef replied:
`sum_overflows` is short and sweet!

koekiebox committed:
Merge branch 'main' into issues/68-merge-commit-transfer

# Conflicts:
#	demos/protobeetle/README.md

koekiebox committed:
Changes include;
* Complete rework of state_machine.zig testing
* Additional assertions
* Review suggestions from DJ/Joran

jorangreef commented on src/state_machine.zig:
```diff
@@ -302,186 +291,278 @@ pub const StateMachine = struct {
+        }
+        if (sum_overflows(t.amount, dr.debits_posted)) return .overflows_debits_posted;
+        if (sum_overflows(t.amount, cr.credits_posted)) return .overflows_credits_posted;
+        if (sum_overflows(dr.debits_pending, dr.debits_posted)) return .overflows_debits;
```
```suggestion
```

This should already be unreachable, since we assert the sum on the previous `create_transfer`. We don't want to return an error for this, but rather have a runtime crash on line 383 if our understanding of this is wrong.

koekiebox replied:
I added this due to a scenario which will not be possible actually. 
```
        mem.zeroInit(Account, .{
            .id = 4,
            .ledger = 1,
            .code = 1,
            .debits_pending = std.math.maxInt(u64) - 1,
            .debits_posted = std.math.maxInt(u64) - 1,
        }),
```

@jorangreef  This will be the case to create the overflow. Do we keep it?

jorangreef replied:
Thanks Jason, we can drop it!

koekiebox replied:
Done.

jorangreef commented on src/state_machine.zig:
```diff
@@ -302,186 +291,278 @@ pub const StateMachine = struct {
+        if (sum_overflows(t.amount, dr.debits_pending + dr.debits_posted)) {
+            return .overflows_debits;
+        }
+        if (sum_overflows(cr.credits_pending, cr.credits_posted)) return .overflows_credits;
```
```suggestion
```

Again, same as before, otherwise it means we somehow accepted a pending transfer in the past that could never be posted.

jorangreef commented on src/state_machine.zig:
```diff
@@ -302,186 +291,278 @@ pub const StateMachine = struct {
+        const insert = self.transfers.getOrPutAssumeCapacity(t.id);
+        if (insert.found_existing) return transfer_exists(t, insert.value_ptr);
 
-    fn create_transfer_rollback(self: *StateMachine, t: Transfer) void {
```
This should stay where it was, right after the `create_transfer()` function. Reason is that it makes it easier to compare the rollback logic, that it's the exact symmetrical inverse of the create.

koekiebox replied:
Fixed.

jorangreef commented on src/state_machine.zig:
```diff
@@ -302,186 +291,278 @@ pub const StateMachine = struct {
-        assert(c.timestamp > self.commit_timestamp);
+    fn post_or_void_pending_transfer(
+        self: *StateMachine,
+        t: *const Transfer,
```
```suggestion
        t: *const Transfer
```

This function signature doesn't need to be wrapped? It should all fit exactly into 100 columns I think?

koekiebox replied:
lol. Yes, on the 100 mark. :-) I thought it went over. Fixed now.

jorangreef commented on src/state_machine.zig:
```diff
@@ -302,186 +291,278 @@ pub const StateMachine = struct {
+            .id = t.id,
+            .debit_account_id = p.debit_account_id,
+            .credit_account_id = p.credit_account_id,
+            .user_data = t.user_data,
```
```suggestion
            .user_data = if (t.user_data > 0) t.user_data else p.user_data,
```

We must inherit the `user_data` from the pending transfer if it isn't defined on the final transfer.

koekiebox replied:
Yes! Good one @jorangreef . I need to add a test for this.

koekiebox replied:
@jorangreef  I added test cases to cover all of the scenarios where posted inherits from pending, including the negative. 
`const f_posted_neg = state_machine.get_transfer(201).?.*;`

jorangreef commented on src/state_machine.zig:
```diff
@@ -302,186 +291,278 @@ pub const StateMachine = struct {
+    fn transfer_exists(t: *const Transfer, e: *const Transfer) CreateTransferResult {
+        assert(t.id == e.id);
+
+        if (t.ledger > 0 and t.ledger != e.ledger) return .exists_with_different_ledger;
```
This should rather stay where it was, i.e. just above the check for `code`, matching the order of the fields of the Transfer struct.

koekiebox replied:
@jorangreef I had to move the logic for the `ledger` check before the `exists_with_different_debit_account_id` / `exists_with_different_credit_account_id`. 
The reason is because you cannot have a transfer where the ledger doesn't match the accounts.
Otherwise we get `transfer_must_have_the_same_ledger_as_accounts`.

jorangreef commented on src/state_machine.zig:
```diff
@@ -302,186 +291,278 @@ pub const StateMachine = struct {
+        if (t.credit_account_id > 0 and t.credit_account_id != e.credit_account_id) {
+            return .exists_with_different_credit_account_id;
+        }
+        if (t.user_data != e.user_data) return .exists_with_different_user_data;
```
```suggestion
        if (t.user_data > 0 and t.user_data != e.user_data) return .exists_with_different_user_data;
```

koekiebox replied:
Done.

jorangreef commented on src/state_machine.zig:
```diff
@@ -302,186 +291,278 @@ pub const StateMachine = struct {
+        if (t.user_data != e.user_data) return .exists_with_different_user_data;
+        if (t.pending_id != e.pending_id) return .exists_with_different_pending_id;
+        if (t.timeout != e.timeout) return .exists_with_different_timeout;
+        if (t.code > 0 and t.code != e.code) return .exists_with_different_code;
```
Great catch! To check if `> 0`.

However, for all these "if present" checks, we should only do that if the transfer is a final transfer, since that's the only case where the fields may be 0.

koekiebox replied:
Agreed. The errors on the initial transfer already ensures that those values are non `0`.

jorangreef commented on src/state_machine.zig:
```diff
@@ -560,151 +649,127 @@ test "create/lookup accounts" {
+            .ledger = 1,
+            .code = 1,
+            .timestamp = 1,
+        }) },
```
```suggestion
        .{ .result = .ok, .object = mem.zeroInit(Account, .{
            .id = 1,
            .ledger = 1,
            .code = 1,
            .timestamp = 1,
        }), },
```

And then `zig fmt` to reflow.

So that the `.result` and `.object` fields are on different lines, following the Zig style guide.

This will also help thin it out a little and make all the `.object` fields line up in the same column, making it easier on the eyes.

koekiebox replied:
Fixed.

jorangreef commented on src/tigerbeetle.zig:
```diff
@@ -78,102 +81,105 @@ pub const Transfer = packed struct {
+    overflows_debits_posted,
+    overflows_credits,
+    overflows_credits_pending,
+    overflows_credits_posted,
```
These should be reordered back to how they were to match the order that they are returned from the state machine.

koekiebox replied:
Fixed.

jorangreef commented on src/tigerbeetle.zig:
```diff
@@ -78,102 +81,105 @@ pub const Transfer = packed struct {
+    pending_transfer_expired,
+
+    pending_transfer_already_posted,
+    pending_transfer_already_voided,
```
These should be reordered to match the order returned from the state machine, i.e. ahead of `pending_transfer_expired`

koekiebox replied:
Fixed.

jorangreef commented on src/tigerbeetle.zig:
```diff
@@ -78,102 +81,105 @@ pub const Transfer = packed struct {
+    pending_transfer_already_posted,
+    pending_transfer_already_voided,
+
+    pending_transfer_has_different_debit_account_id,
+    pending_transfer_has_different_credit_account_id,
+    pending_transfer_has_different_ledger,
+    pending_transfer_has_different_code,
+
+    exceeds_pending_transfer_amount,
```
These should also be reordered to match the order returned from the state machine.

koekiebox replied:
Fixed.

jorangreef reviewed (commented):
Nice work @koekiebox just wanted to get this to you so long.

koekiebox committed:
Changes include;
* Joran review suggestions.

koekiebox committed:
Changes include;
* Add test coverage for data between pending and posted transfers

sentientwaffle commented on src/state_machine.zig:
```diff
@@ -551,7 +623,21 @@ fn equal_48_bytes(a: [48]u8, b: [48]u8) bool {
 const testing = std.testing;
 
-test "create/lookup accounts" {
+test "overflow" {
```
```suggestion
test "sum_overflows" {
```

koekiebox replied:
Done.

koekiebox committed:
Changes include;
* Review comments from Joran
* Review comments from DJ

koekiebox committed:
Changes include:
* Update to tb-node for tb Transfer+Commit change
* Test cases updated
* README.md updated for startup instructions

jorangreef committed:
StateMachine: Fix bugs and improve test coverage

- Fix existing transfers so as not to influence overflow/limit checks.
- Fix create_transfer_exists().
- Fix post_or_void_pending_transfer_exists().
- Fix bug in 2-phase test where f_posted instead of f_posted_neg was used.
- Rework create_transfer() test vectors to also test business logic order.
- Move create_transfer_rollback() to just after create_transfer().
- Assert that the accounts for the pending transfer always exist.
- Use testing.allocator to detect leaks.
- Rename test routines: print_test_vector(), test_account_balances().
- Test more pending/posted balance transitions.
- Do not wrap lines that fit within 100 columns.

jorangreef committed:
StateMachine: Fix order of CreateTransferResult enums

jorangreef committed:
StateMachine: Add create_account_exists() also with more enums, improve coverage

Follow the same create_..._exists() format as with other operations.

Add enums to detect duplicate accounts with different balances:
* exists_with_different_debits_pending
* exists_with_different_debits_posted
* exists_with_different_credits_pending
* exists_with_different_credits_posted

Improve test coverage to detect if any business logic is reordered.
Otherwise, the state machine would no longer be deterministic.

jorangreef committed:
StateMachine: Improve test coverage for two-phase transfers

- Test that business logic is not reordered
- Test rollback for inherited fields correctly
- Test full range of user_data and amount code paths
- Add pending_transfer_has_different_amount for voiding transfer

jorangreef committed:
docs: Update DESIGN doc for new ledger field

jorangreef committed:
docs: Update DESIGN doc description for pending_id

jorangreef commented:
Thanks @koekiebox — landing...
